durable-storage/src/merkle_layer/node.rs

```diff
@@ -0,0 +1,486 @@
+            const RECURSION_LIMIT: u32 = 128;
+            // This function recurses once for every node it traverses, meaning that the number
+            // of recursions are equal to or less than the height of the node.
+            debug_assert!(node.height() <= RECURSION_LIMIT);
+            // The lower bound on the number of nodes in a valid AVL tree is:
+            // fibonacci(height + 3) - 1
+            // see: <https://www.cs.cornell.edu/courses/cs2112/2020fa/lectures/avl/>
+            //
+            // with height = 128:
+            // fibonacci(128 + 3) - 1 > 1x10^27
+            //
+            // This would require:
+            //  - more nodes than 64-bit systems can address.
+            //  - more disk space than has ever been produced.
+            //  - inserting 2 billion nodes every second since the dawn of the universe.
+        }
+        match node.key.cmp(key) {
+            // The key already exists and should be updated.
+            Ordering::Equal => {
+                let node = Arc::make_mut(node);
+                node.key = key.clone();
+                node.data = data.clone();
+                (Some(Arc::new(node.clone())), false)
+            }
+            Ordering::Greater => {
+                let node = Arc::make_mut(node);
+                let (new_left, grew) = set(node.left_mut(), key, data);
+
+                let balance_factor = if grew {
+                    node.balance_factor + 1
+                } else {
+                    node.balance_factor
+                };
+
+                let mut node = Arc::new(MavlNode {
+                    left: new_left,
+                    right: node.right.clone(),
+                    key: node.key.clone(),
+                    data: node.data.clone(),
+                    balance_factor,
+                    ..Default::default()
+                });
+
+                if grew {
+                    let node = rebalance(&mut node);
+                    return (Some(node.clone()), node.balance_factor != 0);
+                }
+
+                #[cfg(debug_assertions)]
+                debug_assert!(
+                    has_correct_balance_factor(&Some(node.clone())),
+                    "Wrong balance factors: {node:?}"
+                );
+
+                (Some(node), false)
+            }
+            Ordering::Less => {
+                let node = Arc::make_mut(node);
+
+                let (new_right, grew) = set(node.right_mut(), key, data);
+
+                let balance_factor = if grew {
+                    node.balance_factor - 1
+                } else {
+                    node.balance_factor
+                };
+
+                let mut node = Arc::new(MavlNode {
+                    left: node.left.clone(),
+                    right: new_right,
+                    key: node.key.clone(),
+                    data: node.data.clone(),
```

Comment:
Are you sure we want to clone the data in this case? That seems like it could be quite expensive.